File: 0008. String to Integer (atoi)/Solution.cpp
#include <iostream>
#include <string>

using std::string;

class Solution {
public:
    int myAtoi(string s) {
        int index = 0;
        bool isnegative = false;
        int result = 0;
        while (s[index] == ' ')
        {
            index++;
        }
        if (s[index] == '-')
        {
            isnegative = true;
            index++;
        }
        else if (s[index] == '+')
        {
            index++;
        }
        while (index < s.size() && s[index] >= '0' && s[index] <= '9')
        {
            int digit = s[index] - '0';
            if (result > INT_MAX / 10 || (result == INT_MAX / 10 && digit > INT_MAX % 10))
                if (isnegative)
                    return INT_MIN;
                else
                    return INT_MAX;
            result = result * 10 + digit;
            index++;
        }
        if (isnegative)
            return result * -1;
        return result;
    }
};

File: 0008. String to Integer (atoi)/Solution.py
class Solution:
    def myAtoi(self, s):
        """
        type s: str
        rtype: int
        """
        result = 0
        index = 0
        negative = False
        s = s.strip()
        if s == "":
            return 0
        if s[index] == '-':
            negative = True
            index += 1
        elif s[index] == '+':
            index += 1
        while index < len(s) and ord(s[index]) >= ord('0') and ord(s[index]) <= ord('9'):
            digit = ord(s[index]) - ord('0')
            if (result > 214748364 or (result == 214748364 and digit > 7)):
                if negative:
                    return -2147483648
                else:
                    return 2147483647
            result = result * 10 + digit
            index += 1
        if (negative):
            return -result
        return result